src/App.jsx

```diff
@@ -14,45 +16,113 @@ export const goodsFromServer = [
-        Reverse
-      </button>
-
-      <button
-        type="button"
-        className="button is-danger is-light"
-      >
-        Reset
-      </button>
-    </div>
+const ALPHABET_SORTING = 'alphabetically';
+const LENGTH_SORTING = 'length';
+const REVERSE = 'reverse';
+const UNREVERSE = 'unreverse';
+const RESET = 'reset';
+let currentGoods = [...goodsFromServer];
+
+export const App = () => {
+  const [currentSorting, setCurentSorting] = useState('');
+
+  const [currentDirection, setCurrentDirection] = useState('');
+
+  if (currentSorting === RESET) {
+    currentGoods = [...goodsFromServer];
+    setCurrentDirection('');
+    setCurentSorting('');
+  }
+
+  if (currentDirection === UNREVERSE) {
+    currentGoods = [...goodsFromServer];
+  }
+
+  if (currentSorting === ALPHABET_SORTING
+     || currentSorting === LENGTH_SORTING) {
+    currentGoods = [...goodsFromServer];
+
+    currentGoods.sort((good1, good2) => {
+      switch (currentSorting) {
+        case ALPHABET_SORTING: {
+          return good1.localeCompare(good2);
+        }
+
+        case LENGTH_SORTING: {
+          return good1.length - good2.length;
+        }
 
-    <ul>
-      <li data-cy="Good">Dumplings</li>
-      <li data-cy="Good">Carrot</li>
-      <li data-cy="Good">Eggs</li>
-      <li data-cy="Good">Ice cream</li>
-      <li data-cy="Good">Apple</li>
-      <li data-cy="Good">...</li>
-    </ul>
-  </div>
-);
+        default: return 0;
+      }
+    });
+  }
+
+  if (currentDirection === REVERSE && currentSorting !== RESET) {
+    currentGoods.reverse();
+  }
+
+  return (
+    <div className="section content">
+      <div className="buttons">
+        <button
+          type="button"
+          className={cn('button',
+            'is-info', {
+              'is-light': currentSorting !== ALPHABET_SORTING,
+            })}
+          onClick={() => setCurentSorting(ALPHABET_SORTING)}
+        >
+          Sort alphabetically
+        </button>
+
+        <button
+          type="button"
+          className={cn('button',
+            'is-success', {
+              'is-light': currentSorting !== LENGTH_SORTING,
+            })}
+          onClick={() => setCurentSorting(LENGTH_SORTING)}
+        >
+          Sort by length
+        </button>
+
+        <button
+          type="button"
+          className={cn('button',
+            'is-warning', {
+              'is-light': currentDirection !== REVERSE,
+            })}
+          onClick={() => {
+            if (currentDirection !== REVERSE) {
+              setCurrentDirection(REVERSE);
+            } else {
+              setCurrentDirection(UNREVERSE);
+            }
+          }
+        }
+        >
+          Reverse
+        </button>
+
+        {currentGoods.some((currentGood, id) => (
+          currentGood !== goodsFromServer[id]
+        )) && (
+        <button
+          type="button"
+          className="button is-danger is-light"
+          onClick={() => setCurentSorting(RESET)}
```

Comment:
```suggestion
          onClick={() => setCurentSorting(RESET)}
```
in reset handler set isReversed to false and currentSorting to NONE
make for it separate function in component and pass reference here.